src/extra/gfx/camera.cpp

```diff
@@ -0,0 +1,187 @@
+#include <foundation.hpp>
+#include <params.hpp>
+#include <linalg_shim.hpp>
+#include <gfx/math.hpp>
+#include <gfx/window.hpp>
+#include <gfx/fmt.hpp>
+#include "../gfx.hpp"
+
+using namespace shards;
+
+namespace gfx {
+
+struct FreeCameraShard : public gfx::BaseConsumer {
+  static SHTypesInfo inputTypes() { return CoreInfo::Float4x4Type; }
+  static SHTypesInfo outputTypes() { return CoreInfo::Float4x4Type; }
+  static SHOptionalString help() { return SHCCSTR("Provides editor free camera controls"); }
+
+  PARAM_PARAMVAR(_flySpeed, "FlySpeed", "Controls fly speed with the keyboard",
+                 {CoreInfo::NoneType, CoreInfo::FloatType, CoreInfo::FloatVarType});
+  PARAM_PARAMVAR(_scrollSpeed, "ScrollSpeed", "Controls middle mouse movement speed",
+                 {CoreInfo::NoneType, CoreInfo::FloatType, CoreInfo::FloatVarType});
+  PARAM_PARAMVAR(_panSpeed, "PanSpeed", "Controls middle mouse pan speed",
+                 {CoreInfo::NoneType, CoreInfo::FloatType, CoreInfo::FloatVarType});
+  PARAM_PARAMVAR(_lookSpeed, "LookSpeed", "Controls right mouse look speed",
+                 {CoreInfo::NoneType, CoreInfo::FloatType, CoreInfo::FloatVarType});
+  PARAM_IMPL(FreeCameraShard, PARAM_IMPL_FOR(_flySpeed), PARAM_IMPL_FOR(_scrollSpeed), PARAM_IMPL_FOR(_panSpeed),
+             PARAM_IMPL_FOR(_lookSpeed));
+
+  struct InputState {
+    struct Keys {
+      bool left{};
+      bool right{};
+      bool forward{};
+      bool backward{};
+    } keys;
+    struct Look {
```

Comment:
All structs and classes with non-trivial behavior should have a comment explaining their function.